src/plonkish/backend/halo2.rs

```diff
@@ -514,3 +679,153 @@ impl<F: PrimeField + From<u64> + Hash> h2Circuit<F> for ChiquitoHalo2SuperCircui
+        Blake2bRead::<_, G1Affine, Challenge255<_>>::init(proof.as_slice());
+    let verifier_params = params.verifier_params();
+    let strategy = SingleStrategy::new(&verifier_params);
+
+    let result = verify_proof_single::<KZGCommitmentScheme<Bn256>, VerifierSHPLONK<Bn256>, _, _, _>(
+        &verifier_params,
+        &vk,
+        strategy,
+        instance,
+        &mut verifier_transcript,
+    );
+    println!("Verify: {:?}", start.elapsed());
+    result
+}
+
+pub fn halo2_prove(
+    params: &ParamsKZG<Bn256>,
+    pk: ProvingKey<G1Affine>,
+    mut rng: BlockRng<OneNg>,
+    cs: ConstraintSystemMid<Fr>,
+    witnesses: Vec<&Assignments<Fr>>,
+    circuits: Vec<ChiquitoHalo2<Fr>>,
+    instance: Vec<Vec<Fr>>,
+) -> Vec<u8> {
+    // Proving
+    println!("Proving...");
+    let start = Instant::now();
+    let mut transcript = Blake2bWrite::<_, G1Affine, Challenge255<_>>::init(vec![]);
+    let engine = PlonkEngineConfig::new()
+        .set_curve::<G1Affine>()
+        .set_msm(H2cEngine::new())
+        .build();
+
+    let mut prover = ProverSingle::<
+            KZGCommitmentScheme<Bn256>,
+            ProverSHPLONK<'_, Bn256>,
+            _,
+            _,
+            _,
+            _,
+        >::new_with_engine(
+            engine,
+            params,
+            &pk,
+            instance.clone(),
+            &mut rng,
+            &mut transcript,
+        )
+        .unwrap();
+
+    for phase in 0..cs.phases() {
+        println!("phase {phase}");
+
+        let mut assigned_witness =
+            vec![Some(vec![Fr::default(); params.n() as usize]); cs.num_advice_columns];
+
+        // TODO as super_circuit.get_mapping().generate() creates a mapping of "compilation unit" ID
+        // to a witness, currently it is impossible to get a witness by the circuit id, hence we are
+        // iterating over all witnesses and checking if the circuit has a column in
+        // question.
+        for witness in witnesses.iter() {
+            for (column, assignments) in witness.iter() {
+                for circuit in circuits.iter() {
+                    if let Some(circuit_column) = circuit.advice_columns.get(&column.uuid()) {
+                        let halo2_column = Column::<Any>::from(*circuit_column);
+                        for (offset, value) in assignments.iter().enumerate() {
+                            assigned_witness[halo2_column.index].as_mut().unwrap()[offset] = *value;
+                        }
+                    }
+                }
+            }
+        }
+
+        // TODO ignoring the challenges produced by the phase, but can they be useful later?
+        let _ = prover.commit_phase(phase as u8, assigned_witness).unwrap();
+        println!("phase {phase} completed");
+    }
+    prover.create_proof().unwrap();
+    let proof = transcript.finalize();
+    println!("Prove: {:?}", start.elapsed());
+
+    proof
+}
+
+#[allow(clippy::type_complexity)]
+pub fn get_halo2_setup(
```

Comment:
I think this could be pub(super) and being called from a method in the result of the plonkish compilation. That method could be added here in a pub trait, so in the future with several packages you don't depend on halo2 if you don't need it.

I think this should return a struct instead of a touple, and to interact with this result calling methods.

Reply:
Perhaps we can see if `k` is something we can calculate from the number of rows in the plonkish compilation.

Reply:
By result of the plonkish compilation do you mean the result of `plonkish()` that is currently a tuple of `(Circuit,Option<AssignmentGenerator>)`?

Reply:
Yes, is it possible to change the result of plonkish::compile to don't be a tuple?

Then add a trait that has a method that returns the halo2 compilation.

Then the return of the halo2 compilation is a struct with methods instead of a long tuble?